src/Nethermind/Nethermind.Evm.Test/Tracing/GasEstimationTests.cs

```diff
@@ -0,0 +1,294 @@
+       {
+           TestEnvironment testEnvironment = new();
+           testEnvironment.CreateEnvironment();
+           long gasLimit = 100000000;
+           Transaction tx = Build.A.Transaction.WithGasLimit(100000000).TestObject;
+           Block block = Build.A.Block.WithNumber(1).WithTransactions(tx).TestObject;
+           
+           EstimateGasTracer tracer = new();
+           long gasLeft = gasLimit - 22000;
+           tracer.ReportAction(gasLeft, 0, Address.Zero, Address.Zero, Array.Empty<byte>(), ExecutionType.Transaction, false);
+           gasLeft = 63 * gasLeft / 64;
+           tracer.ReportAction(gasLeft, 0, Address.Zero, Address.Zero, Array.Empty<byte>(), _executionType, false);
+           gasLeft = 63 * gasLeft / 64;
+           tracer.ReportAction(gasLeft, 0, Address.Zero, Address.Zero, Array.Empty<byte>(), _executionType, false);
+       
+           if (_executionType.IsAnyCreate())
+           {
+               tracer.ReportActionError(EvmExceptionType.Revert, 96000000);
+               tracer.ReportActionError(EvmExceptionType.Revert, 98000000);
+               tracer.ReportActionError(EvmExceptionType.Revert, 99000000);
+           }
+           else
+           {
+               tracer.ReportActionError(EvmExceptionType.Revert, 96000000);
+               tracer.ReportActionError(EvmExceptionType.Revert, 98000000);
+               tracer.ReportActionError(EvmExceptionType.Revert, 99000000);
+           }
+
+           GasEstimator estimator = new(testEnvironment._transactionProcessor, testEnvironment._stateProvider, testEnvironment._specProvider);
+           estimator.Estimate(tx, block.Header, tracer).Should().Be(35146L);
+       }
+       
+       [Test]
+       public void Easy_one_level_case()
+       {
+           TestEnvironment testEnvironment = new();
+           testEnvironment.CreateEnvironment();
+           EstimateGasTracer tracer = new();
+           Transaction tx = Build.A.Transaction.WithGasLimit(128).TestObject;
+           Block block = Build.A.Block.WithNumber(1).WithTransactions(tx).TestObject;
+           
+           tracer.ReportAction(128, 0, Address.Zero, Address.Zero, Array.Empty<byte>(), ExecutionType.Transaction, false);
+           tracer.ReportAction(100, 0, Address.Zero, Address.Zero, Array.Empty<byte>(), _executionType, false);
+       
+           tracer.ReportActionEnd(63, Array.Empty<byte>()); // second level
+           tracer.ReportActionEnd(65, Array.Empty<byte>());
+
+           GasEstimator estimator = new(testEnvironment._transactionProcessor, testEnvironment._stateProvider, testEnvironment._specProvider);
+           estimator.Estimate(tx, block.Header, tracer).Should().Be(1);
+       }
+       
+       [Test]
+       public void Handles_well_nested_calls_where_most_nested_defines_excess()
+       {
+           TestEnvironment testEnvironment = new();
+           testEnvironment.CreateEnvironment();
+           EstimateGasTracer tracer = new();
```

Comment:
1. make creating TestEnvironment not need to call CreateEnvironment
2. Encapsulate EstimateGasTracer and GasEstimator creation into TestEnvironment